docs/src/content/docs/book/assembly-functions.mdx

```diff
@@ -0,0 +1,678 @@
+
+```tact
+// Pushes `x` onto the stack, increments it there,
+// but does not capture the result, leaving it on the stack
+asm fun push(x: Int) { INC }
+```
+
+Specifying a [primitive type][p], such as an [`Int{:tact}`][int] or a [`Cell{:tact}`][cell], will make the assembly function pop the top value from the stack and produce it as a result. If the run-time type of the popped value doesn't match the specified return type, an exception with [exit code 7](/book/exit-codes#7) will be thrown: `Type check error`.
+
+```tact
+// CAUSES RUN-TIME ERROR!
+// Pushes `x` onto the stack, then tries to capture it as a Cell,
+// causing an exit code 7: Type check error
+asm fun push(x: Int): Cell { }
+```
+
+Just like in [parameters](#conventions-parameters), arbitrary [Struct][struct] return types are distributed across their fields and recursively flattened in exactly the same order. The only differences are that they now capture or pop values from the stack instead of pushing them onto the stack, and they do so in a right-to-left fashion — the last field of the [Struct][struct] pops the topmost value from the stack, the second-to-last pops the second to the top, and so on, so that the last field contains the value from the top of the stack and the first field contains the value from the bottom.
+
+```tact
+// Struct with two fields of type Int
+struct MinMax { minVal: Int; maxVal: Int }
+
+// Pushes `a` and `b` onto the stack,
+// then captures two values back via the `MinMax` Struct
+asm fun minmax(a: Int, b: Int): MinMax { MINMAX }
+```
+
+If the run-time type of some popped value doesn't match some specified field type of the [Struct][struct] or the nested [Structs][struct], if any, an exception with [exit code 7](/book/exit-codes#7) will be thrown: `Type check error`. Moreover, attempts to capture more values than there were on the stack throw an exception with [exit code 2](/book/exit-codes#2): `Stack underflow`.
+
+```tact
+// Struct with way too many fields for initial stack to handle
+struct Handler { f1: Int; f2: Int; f3: Int; f4: Int; f5: Int; f6: Int; f7: Int }
+
+// CAUSES RUN-TIME ERROR!
+// Tries to capture 7 values from the stack and map them onto the fields of `Handler`,
+// but there's just isn't that many values on the initial stack after TVM initialization,
+// which causes an exit code 2 to be thrown: Stack underflow
+asm fun overHandler(): Handler { }
+```
+
+As parameters and return values of assembly functions, [Structs][struct] can only have up to $16$ fields on each nested level. That is, you could go over $16$ fields in total if, for example, some of them were the nested structures, thus increasing the nesting level. Note that the outer parameters of the function also count as a nested level, so specifying a [Struct][struct] with more than $16$ fields as parameters or with more than $16$ parameters will cause compilation errors, especially in assembly functions.
```

Comment:
> So, the following is OK because I do not exceed 16 fields (and function params) at each nested level? ...code block...

Yes

> But I will get a compilation error if I do either of the following actions:
> 
> Add an extra field to struct A, because there would be more than 16 fields in struct A.
> Add an extra field to struct B, because there would be more than 16 fields in struct B.
> Add an extra parameter to function test, because there would be more than 16 parameters in test (function parameters act as the outermost nested level).

Yes

> I wonder if we could avoid the concept of nested levels in the explanation and say:
> <<As parameters and return values of assembly functions, structs can only have up to 
> 16
>  fields. Each nested struct in those fields can also have up to 16 fields. This restriction also applies to asm function parameters, which can only have up to 16 parameters.>>

Hmm, maybe

> Also, I did not get what "especially in assembly functions" means.

"especially" there probably has to be replaced by "only" — the intent was to say that there are no such limitation in non-asm functions, but only because of #590. And, on that note, the Structs are encoded different in different places of Tact, which isn't what we want to document (probably), at least until we sort it out with #1127, #458 and related.